### PR TITLE
Allow admins to pre-configure desktop apps

### PR DESCRIPTION
**Pull request by yuya-oc**
Before submitting, please confirm you've
 - [x] read and understood our [Contributing Guidelines](https://github.com/mattermost/desktop/blob/master/CONTRIBUTING.md)
 - [x] completed [Mattermost Contributor Agreement](http://www.mattermost.org/mattermost-contributor-agreement/)
 - [x] executed `npm run lint:js` for proper code formatting

Please provide the following information:

**Summary**
Allow admins to pre-configure desktop apps.

By creating the `preconfig.json`, admins can set defaultPreferences for the desktop app.
For now, auto-start is not configured by this PR yet.

**Issue link**
#483

**Test Cases**
1. Quit the app then remove existing `config.json`.
2. Create `preconfig.json` at following folder.
    - Windows: C:\ProgramData\Mattermost\
    - Mac: /Library/Application Support/Mattermost/
    - Linux: /etc/mattermost/
3. `preconfig.json` should have following structure.
    ```
    {
      "defaultPreferences": {
        "version": 1,
        "teams": [
          ...
        ],
        ...(same keys/values as config.json)
      }
    }
    ```
4. The app should start with pre-configured defaultPreferences.

**Additional Notes**
https://circleci.com/gh/yuya-oc/desktop/519#artifacts

### CODE REVIEW

**Comment by jasonblais** [comment]
This is great, yes! I'll test this out

**Comment by yuya-oc** [comment]
https://circleci.com/gh/yuya-oc/desktop/525#artifacts

Now `autoStart` can be configured by preconfig.json

File: src/common/config/defaultPreferences.js [hunk 1]
```diff
@@ -15,7 +15,8 @@ const defaultPreferences = {
   },
   showUnreadBadge: true,
   useSpellChecker: true,
-  spellCheckerLocale: 'en-US'
+  spellCheckerLocale: 'en-US',
+  autoStart: false
```
Comment (jasonblais) [inline src/common/config/defaultPreferences.js]: Currently, `autoStart` is `true`, I believe?

Comment (yuya-oc) [reply src/common/config/defaultPreferences.js]: I have thought the default is false though, the document says it's true.
https://docs.mattermost.com/help/apps/desktop-guide.html#app-options

Should we enable this?

**Comment by jasonblais** [comment]
For Mac, I tried the following:

1. Install the artifact. Open it to confirm it's working properly. Quit the app then remove existing `config.json`.
2. Create `preconfig.json` at following folder - Mac: /Library/Application Support/Mattermost/
```
{
  "version": 1,
  "teams": [
    {
      "name": "Pre-Release",
      "url": "https://pre-release.mattermost.com"
    },
    {
      "url": "https://example.com",
      "name": "Example",
      "index": false
    },
    {
      "name": "Spinmint",
      "url": "http://i-07cfb2731385c7fd0.spinmint.com/admin_console/advanced/message_export"
    }
  ],
  "showTrayIcon": false,
  "trayIconTheme": "light",
  "minimizeToTray": false,
  "notifications": {
    "flashWindow": 0,
    "bounceIcon": false,
    "bounceIconType": "informational"
  },
  "showUnreadBadge": true,
  "useSpellChecker": false,
  "spellCheckerLocale": "en-US",
  "autoStart": false
}
```
3. Open the app from the Downloads folder

Observed: No config.json file created - all settings are defaults.

**Comment by jasonblais** [comment]
I'll test Windows next,

File: src/common/config/defaultPreferences.js (hunk 1 above)
Comment (jasonblais) [reply src/common/config/defaultPreferences.js]: I think we should keep the default setting at what it's now. If the default is `false` in the code, let's keep it `false` and I can just update the documentation.

**Comment by yuya-oc** [comment]
For now, config.json is not created until the settings are changed in the settings page.

File: src/common/config/defaultPreferences.js (hunk 1 above)
Comment (yuya-oc) [reply src/common/config/defaultPreferences.js]: I reinstalled v3.7.1 after removing config.json. The default is false.

**Comment by jasonblais** [comment]
So what is the expected on step 3? As a user, I just saw empty settings fields with default values.

File: src/common/config/defaultPreferences.js (hunk 1 above)
Comment (jasonblais) [reply src/common/config/defaultPreferences.js]: Sounds good, let's keep it false. Updated doc https://github.com/mattermost/docs/commit/5892490d7ce36e80acac7fdfdfada261be5b5e56

**Comment by yuya-oc** [comment]
"No config.json" is expected. By the way, I think `defaultPreferences` key is missing in `preconfig.json`, so it should be below:

```json
{
  "defaultPreferences": {
    "version": 1,
    "teams": [
      {
        "name": "Pre-Release",
        "url": "https://pre-release.mattermost.com"
      },
      {
        "url": "https://example.com",
        "name": "Example",
        "index": false
      },
      {
        "name": "Spinmint",
        "url": "http://i-07cfb2731385c7fd0.spinmint.com/admin_console/advanced/message_export"
      }
    ],
    "showTrayIcon": false,
    "trayIconTheme": "light",
    "minimizeToTray": false,
    "notifications": {
      "flashWindow": 0,
      "bounceIcon": false,
      "bounceIconType": "informational"
    },
    "showUnreadBadge": true,
    "useSpellChecker": false,
    "spellCheckerLocale": "en-US",
    "autoStart": false
  }
}
```

**Comment by yuya-oc** [comment]
Rebased against to the latest master. Artifacts will be available at https://circleci.com/gh/yuya-oc/desktop/526#artifacts

**Comment by jasonblais** [comment]
@yuya-oc I used the updated file you provided, but I can't seem to get it to work. Not sure why.

On Windows, I don't have a Mattermost folder in `C:\ProgramData. So I created one, added the preconfig.json file there and unfortunately it didn't work.

Let me know if I should give more information to try and troubleshoot the issue?

**Comment by yuya-oc** [comment]
I got following config by using the `C:\ProgramData\Mattermost\preconfig.json`. So I think the feature is working.

mattermost_preconfig

@jason Would you make sure whether the `preconfig.json` can be opened with other apps (e.g. notepad.exe)?

**Comment by jasonblais** [comment]
@yuya-oc I did get an administrator permissions error at one point on Windows when I tried to create it. Could that be why?

**Comment by yuya-oc** [comment]
I think that's not matter. The C:\ProgramData is originally system folder, so Windows requested a permission to write. However if you can read the json with notepad.exe, you should have read-only permission at least. That means that the desktop app should be able to read the json, too.

**Comment by jasonblais** [comment]
@yuya-oc I verified I can open the file on Notepad and Notepad++

**Comment by yuya-oc** [comment]
@jasonblais Hmm...it's weird. Assuming that, I can only think that the file name is not correct.

**Comment by yuya-oc** [comment]
Looks correct to me. And the json content is exactly same as mine.

**Comment by yuya-oc** [comment]
Added some log messages: https://circleci.com/gh/yuya-oc/desktop/527#artifacts

**Comment by jasonblais** [comment]
@yuya-oc After removing the AppData/Local and AppData/Roaming folders, this worked fine on Windows 10.

Assuming that's expected, this looks good.

Should I test this on Mac as well?

**Comment by yuya-oc** [comment]
@jasonblais Thanks! Yeah, we should test on Mac because the preconfig.json path is different from Windows.

**Comment by jasonblais** [comment]
@yuya-oc That makes sense.

On Mac, do I have to delete a folder, similar to Windows? Because on Windows it didn't work by simply removing config.json, I had to delete AppData/Local and/or AppData/Roaming folders as well.

Which leads to my second question - if the end user has ever installed the Mattermost app before, getting a pre-configured app from their admin won't work unless they remove the appropriate files/folders. Is that correct?

**Comment by jasonblais** [comment]
@yuya-oc 

> About config.json, yes.

Is this is the only option?

**Comment by jasonblais** [comment]
@yuya-oc And for you, it works by simply deleting `config.json`?

**Comment by yuya-oc** [comment]
>> About config.json, yes.

> Is it the only option?

For now, yes. To solve this problem, we need to consider about priority of three configurations (config, preconfig, buildConfig). Or, need to create new mechanism to specify priority for each parameter.

> And for you, it works by simply deleting config.json?

Yes. I didn't remove AppData/Roaming folder while developing.

**Comment by jasonblais** [comment]
> To solve this problem, we need to consider about priority of three configurations (config, preconfig, buildConfig). Or, need to create new mechanism to specify priority for each parameter.

If we merge this PR, is it going to make it harder to solve this problem? Why not resolve this with this PR, or is that going to be a lot of work to create the new mechanism?

**Comment by jasonblais** [comment]
@yuya-oc When I test on Mac, I end up with the same result as described in https://github.com/mattermost/desktop/pull/672#issuecomment-349329809

I downloaded the latest artifact from https://circleci.com/gh/yuya-oc/desktop/527#artifacts, so I can share the log messages you added. Let me know where it's stored in Mac.

**Comment by yuya-oc** [comment]
@jasonblais In Terminal.app, please execute following command.

```
/PATH/TO/Mattermost.app/Contents/MacOS/Mattermost
```

Logs would appear on the terminal window.

**Comment by yuya-oc** [comment]
> If we merge this PR, is it going to make it harder to solve this problem? Why not resolve this with this PR, or is that going to be a lot of work to create the new mechanism?

Perhaps a little hard (a few days?). Currently we are going to allow any parameters in preconfig though, I feel it makes confusing between user's preferences and admin's preferences.
For example, probably admins would want to pre-configure server URLs. On the other hand, users would want to configure behavior preferences for themselves such as auto-start, badges, notifications, etc.

But in current mechanism, we can only prioritize one of config and preconfig like [`deepmerge`](https://github.com/KyleAMathews/deepmerge) because we equally handle parameters. Possibly, to restrict parameters which can be pre-configured might make sense.

**Comment by jasonblais** [comment]
@yuya-oc This is a silly question..but what is a typical path for `/PATH/TO`? I'm new to Mac..

**Comment by jasonblais** [comment]
Also, sorry for the delays, had some travelling last two days

**Comment by yuya-oc** [comment]
@jasonblais In general use case, Mac apps should be located under `/Applications`. (homebrew-cask also do that)

**Comment by amyblais** [comment]
@yuya-oc I tested this in Windows and could not get it to work - I'm able to get the server management page open but it doesn't have any preconfigured info - please see attached screenshot:

screenshot_14

I didn't delete AppData/Local and/or AppData/Roaming folders as Jason had done because I thought it should work without having to do this.

**Comment by jasonblais** [comment]
@yuya-oc Any thoughts? I think this is the same issue I was having until I deleted the folders in AppData.

**Comment by yuya-oc** [comment]
Assuming that, probably there is significant reason. But still I’m not sure.

@jason What about https://github.com/mattermost/desktop/pull/672#issuecomment-352033686 log message?

**Comment by jasonblais** [comment]
@yuya-oc I'll try to re-test this next week

**Comment by jasonblais** [comment]
@yuya-oc Terribly sorry it took me this long.

Here are the logs from Mac. When deleting config.json, it cannot find the file, and doesn't look for preconfig.json

```
Failed to read or upgrade config.json { Error: ENOENT: no such file or directory, open '/Users/jasonblais/Library/Application Support/Mattermost/config.json'
    at Object.fs.openSync (fs.js:584:18)
    at Object.module.(anonymous function) [as openSync] (ELECTRON_ASAR.js:173:20)
    at Object.fs.readFileSync (fs.js:491:33)
    at Object.fs.readFileSync (ELECTRON_ASAR.js:505:29)
    at Object.readFileSync (/Applications/Mattermost.app/Contents/Resources/app.asar/main_bundle.js:1160:34)
    at Object.<anonymous> (/Applications/Mattermost.app/Contents/Resources/app.asar/main_bundle.js:7679:21)
    at __webpack_require__ (/Applications/Mattermost.app/Contents/Resources/app.asar/main_bundle.js:20:30)
    at /Applications/Mattermost.app/Contents/Resources/app.asar/main_bundle.js:66:18
    at Object.<anonymous> (/Applications/Mattermost.app/Contents/Resources/app.asar/main_bundle.js:69:10)
    at Object.<anonymous> (/Applications/Mattermost.app/Contents/Resources/app.asar/main_bundle.js:18575:3)
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/Users/jasonblais/Library/Application Support/Mattermost/config.json' }

```

**Comment by jasonblais** [comment]
We should also think about the previous note about the priority of config, preconfig, and buildConfig.

https://github.com/mattermost/desktop/pull/672#issuecomment-351437840

**Comment by jasonblais** [comment]
Hey @yuya-oc - any questions about the above?

**Comment by jasonblais** [comment]
@yuya-oc @esethna @wget I'm wondering if this PR is still useful or if the work done here is covered by other admin configuration improvements?

**Comment by hanzei** [comment]
@deanwhillier What do you think about the questions from Jason above?

**Comment by hanzei** [comment]
@deanwhillier Gentle reminder about the above question.

**Comment by Willyfrog** [comment]
closing as this is covered by several other means, but will gladly reopen if there is anything missing :)

**Comment by ompel** [comment]
Sorry to resurrect PRs from the dead, but can't we simply have an easier way to preconfigure the desktop apps? Building and packaging the source code just to preconfigure the server URL seems like an incredible amount of work for something that could be easily achieved with the implementation suggested in this PR. Not to mention, for signing a Mac app, you'd also have to set up an Apple developer account, etc.

Or perhaps I'm overlooking something in the documentation?

**Comment by jyrkij** [comment]
> closing as this is covered by several other means, but will gladly reopen if there is anything missing :)

What are these 'several other means'?  https://mattermost.atlassian.net/browse/MM-14086 is closed as invalid, because

> This is now being handled by GPO configuration options with the MSI installer.

This doesn't complete the use case as MSI/GPO is only for Windows.  How to accomplish this on macOS or Linux without building from source code?

**Pull request closed without merging**